File: scripts/run_batch.py
import os
import subprocess
from pathlib import Path

PROJECT_ROOT = Path(__file__).resolve().parent.parent

PIN = PROJECT_ROOT / "pin_kit/pin"

PINTOOL = (
    PROJECT_ROOT /
    "pin_kit/source/tools/MyPinTool/obj-intel64/MyPinTool.so"
)

TESTS_DIR = PROJECT_ROOT / "tests"
BIN_DIR = TESTS_DIR / "bin"
REPORTS_DIR = PROJECT_ROOT / "reports"

BIN_DIR.mkdir(exist_ok=True)
REPORTS_DIR.mkdir(exist_ok=True)

cpp_files = list(TESTS_DIR.glob("*.cpp"))

if not cpp_files:
    print("No test files found.")
    exit(0)

print(f"Found {len(cpp_files)} test files.\n")

for cpp_file in cpp_files:
    test_name = cpp_file.stem
    binary_path = BIN_DIR / test_name
    report_path = REPORTS_DIR / f"{test_name}.json"

    print("=" * 60)
    print(f"TEST: {test_name}")
    print("=" * 60)

    compile_cmd = [
        "g++",
        str(cpp_file),
        "-o",
        str(binary_path)
    ]

    print("\n[1] Compiling...")

    compile_result = subprocess.run(
        compile_cmd,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        text=True
    )

    if compile_result.returncode != 0:
        print("Compilation FAILED\n")
        print(compile_result.stderr)
        continue

    print("Compilation SUCCESS")

    pin_cmd = [
        str(PIN),
        "-t",
        str(PINTOOL),
        "-o",
        str(report_path),
        "--",
        str(binary_path)
    ]

    print("\n[2] Running Pintool...")

    run_result = subprocess.run(
        pin_cmd,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        text=True
    )

    if run_result.returncode != 0:
        print("Pintool execution FAILED\n")
        print(run_result.stderr)
        continue

    print("Pintool SUCCESS")

    print(f"\nGenerated report:")
    print(report_path)

print("\n")
print("=" * 60)
print("BATCH TESTING COMPLETE")
print("=" * 60)

File: scripts/run_batch.js
const fs = require('fs');
const path = require('path');
const { spawnSync } = require('child_process');

const PROJECT_ROOT = path.resolve(__dirname, '..');

const PIN = path.join(PROJECT_ROOT, 'pin_kit', 'pin');
const PINTOOL = path.join(
  PROJECT_ROOT,
  'pin_kit',
  'source',
  'tools',
  'MyPinTool',
  'obj-intel64',
  'MyPinTool.so',
);

const TESTS_DIR = path.join(PROJECT_ROOT, 'tests');
const BIN_DIR = path.join(TESTS_DIR, 'bin');
const REPORTS_DIR = path.join(PROJECT_ROOT, 'reports');

if (!fs.existsSync(BIN_DIR)) {
  fs.mkdirSync(BIN_DIR, { recursive: true });
}
if (!fs.existsSync(REPORTS_DIR)) {
  fs.mkdirSync(REPORTS_DIR, { recursive: true });
}

const cppFiles = fs
  .readdirSync(TESTS_DIR)
  .filter((file) => file.endsWith('.cpp'))
  .map((file) => path.join(TESTS_DIR, file));

if (cppFiles.length === 0) {
  console.log('No test files found.');
  process.exit(0);
}

console.log(`Found ${cppFiles.length} test files.\n`);

for (const cppFile of cppFiles) {
  const testName = path.parse(cppFile).name;
  const binaryPath = path.join(BIN_DIR, testName);
  const reportPath = path.join(REPORTS_DIR, `${testName}.json`);

  console.log('='.repeat(60));
  console.log(`TEST: ${testName}`);
  console.log('='.repeat(60));

  console.log('\n[1] Compiling...');
  const compileArgs = ['-g', cppFile, '-o', binaryPath];

  const compileResult = spawnSync('g++', compileArgs, { encoding: 'utf-8' });

  if (compileResult.status !== 0) {
    console.log('Compilation FAILED\n');
    console.error(compileResult.stderr);
    continue;
  }
  console.log('Compilation SUCCESS');

  console.log('\n[2] Running Pintool...');
  const pinArgs = ['-t', PINTOOL, '-o', reportPath, '--', binaryPath];

  const runResult = spawnSync(PIN, pinArgs, { encoding: 'utf-8' });

  if (runResult.status !== 0) {
    console.log('Pintool execution FAILED\n');
    console.error(runResult.stderr);
    continue;
  }

  console.log('Pintool SUCCESS');
  console.log(`\nGenerated report:\n${reportPath}`);
}

console.log('\n' + '='.repeat(60));
console.log('BATCH TESTING COMPLETE');
console.log('='.repeat(60));
